apollo-router/src/plugins/telemetry/config_new/instruments.rs

```diff
@@ -1545,17 +1722,80 @@ where
         attrs.append(&mut inner.attributes);
 
         let increment = match inner.increment {
-            Increment::Unit | Increment::EventUnit => Some(1f64),
+            Increment::Unit | Increment::EventUnit | Increment::FieldUnit => Some(1f64),
             Increment::Duration(instant) | Increment::EventDuration(instant) => {
                 Some(instant.elapsed().as_secs_f64())
             }
-            Increment::Custom(val) | Increment::EventCustom(val) => val.map(|incr| incr as f64),
+            Increment::Custom(val) | Increment::EventCustom(val) | Increment::FieldCustom(val) => {
+                val.map(|incr| incr as f64)
+            }
         };
 
         if let (Some(histogram), Some(increment)) = (inner.histogram.take(), increment) {
             histogram.record(increment, &attrs);
         }
     }
+
+    fn on_response_field(&self, typed_value: &TypedValue, ctx: &Context) {
+        let mut inner = self.inner.lock();
+        if !inner.condition.evaluate_response_field(typed_value, ctx) {
+            return;
+        }
+
+        // Response field may be called multiple times so we don't extend inner.attributes
+        let mut attrs = inner.attributes.clone();
+        if let Some(selectors) = inner.selectors.as_ref() {
+            attrs.extend(
+                selectors
+                    .on_response_field(typed_value, ctx)
+                    .into_iter()
+                    .collect::<Vec<_>>(),
+            );
+        }
+
+        if let Some(selected_value) = inner
+            .selector
+            .as_ref()
+            .and_then(|s| s.on_response_field(typed_value, ctx))
+        {
+            let new_incr = match &inner.increment {
+                Increment::FieldCustom(None) => {
+                    Increment::FieldCustom(selected_value.as_str().parse::<i64>().ok())
+                }
+                Increment::Custom(None) => {
+                    Increment::FieldCustom(selected_value.as_str().parse::<i64>().ok())
+                }
+                other => {
+                    failfast_error!("this is a bug and should not happen, the increment should only be Custom or FieldCustom, please open an issue: {other:?}");
```

Comment:
I don't think I've come across this macro yet. Are we ever actually compiling with the failfast flag that causes this to panic?

Reply:
No, it's just for testing when developing locally, I would probably be in favour of removing this macro in general in a separate PR.